### PR TITLE
Move logger and profiler finalization to trainer's teardown

### PR DESCRIPTION
**Pull request by tangbinh**
## What does this PR do?

Previously, profiler finalization was implemented in multiple places, including the evaluation loop, prediction loop, and fit loop. Logger finalization, however, was included in the fit loop only, leading to Issue #8333.

Here, we move the logger and profiler finalization to the `_call_teardown_hook` in the trainer and call the hook from each process in case of DDP or TPU training. See #8217 for more contexts and discussions (cc @ananthsub).

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 1]
```diff
@@ -203,6 +203,9 @@ def new_process(self, process_idx, trainer, mp_queue):
         # persist info in ddp_spawn
         self.transfer_distrib_spawn_state_on_fit_end(results)
 
+        # https://github.com/PyTorchLightning/pytorch-lightning/pull/8217#issuecomment-871261949
+        trainer._call_teardown_hook(self.lightning_module)
+
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:208]: Please add some text comments in addition to the link to the GitHub issue. it'll be easier to read rather than needing ro follow links

something like
```
# Ensure that spawned processes go through teardown before joining
# for more context, see <GitHub link> 
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1208,7 +1214,7 @@ def _call_teardown_hook(self, model: "pl.LightningModule") -> None:
 
         if self.datamodule is not None:
             self.datamodule.teardown(stage=fn)
-        self.profiler.teardown(stage=fn)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1211]: where is the profiler teardown happening now?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -943,7 +944,12 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
             self.call_hook("on_fit_end")
 
         # teardown
-        self._call_teardown_hook(model)
+        if self._distrib_type not in (
+            DistributedType.DDP_SPAWN,
+            DistributedType.TPU_SPAWN,
+            DistributedType.DDP_SHARDED_SPAWN,
+        ):
+            self._call_teardown_hook(model)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:952]: same here: please add a comment for why spawn plugins are special-cased
@awaelchli @kaushikb11  - should we have a private helper function for determining if checking if the training type is spawned? would this be useful elsewhere?

**Review by ananthsub — COMMENTED** [review]
be sure to update the changelog as well

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py:1211]: It appears that `self.profiler.describe()` already includes a teardown call ([see this line](https://github.com/PyTorchLightning/pytorch-lightning/blob/d83dd7969db1857daac28ec352c177f41d4583eb/pytorch_lightning/profiler/base.py#L153)), so I think it is not necessary to invoke another one.

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1211]: ah right. that's something we'll clear up separately outside of this PR

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -173,6 +173,14 @@ def new_process(self, process_idx: int, trainer, mp_queue) -> None:
         if self.local_rank == 0:
             time.sleep(2)
 
+<<<<<<< HEAD
+        # ensure that spawned processes go through teardown before joining (for more context, see
+        # https://github.com/PyTorchLightning/pytorch-lightning/pull/8217#issuecomment-871261949).
+=======
+        # https://github.com/PyTorchLightning/pytorch-lightning/pull/8217#issuecomment-871261949
+>>>>>>> c19981214f0d55516571570d991a21e16f944c98
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:181]: address merge conflicts

**Review by ananthsub — COMMENTED** [review]
please update the changelog with this fix as well: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md

**Review by ananthsub — COMMENTED** [review]
@tangbinh - could you also add a test that confirms that the logger is finalized when calling trainer.fit, trainer.validate, trainer.test, and trainer.predict? 

you can see tests/trainer/logging_/test_distributed_logging.py & https://github.com/PyTorchLightning/pytorch-lightning/pull/8608/files for examples of integration tests between the trainer and loggers. 

you can create a dummy logger or mock and confirm that finalize is called

File: pytorch_lightning/loops/fit_loop.py [hunk 5]
```diff
@@ -225,15 +225,6 @@ def on_run_end(self) -> None:
...
 
-        # todo: TPU 8 cores hangs in flush with TensorBoard. Might do for all loggers.
-        # It might be related to xla tensors blocked when moving the cpu
-        # kill loggers
-        if self.trainer.logger is not None:
-            self.trainer.logger.finalize("success")
-
-        # summarize profile results
-        self.trainer.profiler.describe()
-
```
Comment (ananthsub) [inline pytorch_lightning/loops/fit_loop.py:236]: these need to be added to the trainer's teardown hook

Comment (tangbinh) [reply pytorch_lightning/loops/fit_loop.py:236]: My bad! I was merging the commits and somehow missed these changes. Thank you for spotting it!

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:953]: You can add a staticmethod like `interactive_compatible_types`

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:953]: can we do an instance check against the spawn plugin from which all these other plugins inherit?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 6]
```diff
@@ -203,6 +203,10 @@ def new_process(self, process_idx, trainer, mp_queue):
         # persist info in ddp_spawn
         self.transfer_distrib_spawn_state_on_fit_end(results)
 
+        # ensure that spawned processes go through teardown before joining (for more context, see
+        # https://github.com/PyTorchLightning/pytorch-lightning/pull/8217#issuecomment-871261949).
+        trainer._call_teardown_hook(self.lightning_module)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:208]: Thanks for thinking about my comment here and including it. 

However, I think it would be better to keep a small comment here and either point to the full pull request or not put a link at all. While the contents of my comments are correct, the formulation is not very formal to be included in the source could I believe.

**Review by awaelchli — COMMENTED** [review]
overall LGTM :) thanks for working on it

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -942,8 +943,14 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
...
 
-        # teardown
-        self._call_teardown_hook(model)
+        # teardown if necessary (similar calls for spawn plugins are excluded as they have
+        # been included at the end of `new_process` functions)
+        if self._distrib_type not in (
+            DistributedType.DDP_SPAWN,
+            DistributedType.TPU_SPAWN,
+            DistributedType.DDP_SHARDED_SPAWN,
+        ):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:952]:
```suggestion
        # teardown if necessary (similar calls for spawn plugins are excluded as they have
        # been included at the end of `new_process` functions)
        if not self._distrib_type.is_interactive_compatible():
```
https://github.com/PyTorchLightning/pytorch-lightning/blob/9ca02f58aeba6427f1a45b18818307e95b3f4ac1/pytorch_lightning/utilities/enums.py#L73-L76

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:953]: @awaelchli @kaushikb11 this could be a good follow up. the usage of these enums (device type, distributed type, amp, grad clip algorithm) will run into issues if/when users supply their own implementations. instance checks will enforce an inheritance structure for users to follow. i wonder if explicit properties per accelerator strategy plugin is another option

Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py]: Interesting.

For eg., a User implements their own `tpu_sharded_spawn` plugin. It would cause issues to the User, should we have functionality for Users to register or update attributes in Enums?

Yes, having explicit properties for accelerator strategy plugins is one option, but I like the idea of having it decoupled in Enums rather than having it with plugins.

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -942,8 +943,10 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
         if self.state.fn == TrainerFn.FITTING:
             self.call_hook("on_fit_end")
 
-        # teardown
-        self._call_teardown_hook(model)
+        # teardown if necessary (similar calls for spawn plugins are excluded as they have
+        # been included at the end of `new_process` functions)
+        if not self._distrib_type.is_interactive_compatible():
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:948]:
```suggestion
        if self._distrib_type is None or not self._distrib_type.is_interactive_compatible():
```

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (tangbinh) [reply pytorch_lightning/trainer/trainer.py]: @ananthsub @kaushikb11 Those are good points. In this case, is it better to replace the enum check with `isinstance(self.training_type_plugin, ParallelPlugin)` as suggested by @awaelchli then?

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -942,8 +943,10 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
         if self.state.fn == TrainerFn.FITTING:
             self.call_hook("on_fit_end")
 
-        # teardown
-        self._call_teardown_hook(model)
+        # teardown if necessary (similar calls for spawn plugins are excluded as they have
+        # been included at the end of `new_process` functions)
+        if self._distrib_type is None or not self._distrib_type.is_interactive_compatible():
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:948]: for a followup, the typehint here is wrong: https://github.com/PyTorchLightning/pytorch-lightning/blob/9ca02f58aeba6427f1a45b18818307e95b3f4ac1/pytorch_lightning/trainer/properties.py#L121-L123

It should be Optional[DistributedType]. https://github.com/PyTorchLightning/pytorch-lightning/blob/9ca02f58aeba6427f1a45b18818307e95b3f4ac1/pytorch_lightning/trainer/connectors/accelerator_connector.py#L106

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:948]:
```suggestion
        if self._distrib_type not in DistributedType.interactive_compatible_types():
```

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py]: for now, i think @carmocca 's suggestion for `interactive_compatible_types` is good to proceed with this fix.
we can file a separate issue to audit the lightning codebase for how/where DistributedType is used

**Review by ananthsub — APPROVED** [review]
LGTM, but please address @awaelchli 's comment on comments as well

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -942,8 +943,10 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
         if self.state.fn == TrainerFn.FITTING:
             self.call_hook("on_fit_end")
 
-        # teardown
-        self._call_teardown_hook(model)
+        # teardown if necessary (similar calls for spawn plugins are excluded as they have
+        # been included at the end of `new_process` functions)
+        if self._distrib_type not in DistributedType.interactive_compatible_types():
+            self._call_teardown_hook(model)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:949]: I will follow up on this. I am convinced `DistributedType.interactive_compatible_types` is not the right choice for what we are trying to solve here as it contains too many plugins.

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 11]
```diff
@@ -102,6 +102,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     [#8627](https://github.com/PyTorchLightning/pytorch-lightning/pull/8627))
 
 
+- Fix an issue with logger ouputs not being finalized correctly after prediction runs ([#8333](https://github.com/PyTorchLightning/pytorch-lightning/issues/8333))
```
Comment (awaelchli) [inline CHANGELOG.md:105]:
```suggestion
- Fixed an issue with logger outputs not being finalized correctly after prediction runs ([#8333](https://github.com/PyTorchLightning/pytorch-lightning/issues/8333))
```

**Review by carmocca — APPROVED** [review]
